src/System.Transactions/tests/IntResourceManager.cs

```diff
@@ -0,0 +1,254 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Xunit;
+
+namespace System.Transactions.Tests
+{
+    // Ported from Mono
+
+    public class IntResourceManager
+    {
+        public IntResourceManager(int value)
+        {
+            _actual = value;
+            _guid = Guid.NewGuid();
+        }
+
+        private int _actual;
+        private int _tmpValue;
+        private Transaction _transaction = null;
+
+        public int NumPrepare = 0;
+        public int NumRollback = 0;
+        public int NumCommit = 0;
+        public int NumInDoubt = 0;
+        public int NumSingle = 0;
+
+        public int NumInitialize = 0;
+        public int NumPromote = 0;
+        public int NumEnlistFailed = 0;
+
+        public ResourceManagerType Type = ResourceManagerType.Volatile;
+        public bool FailPrepare = false;
+        public bool FailWithException = false;
+        public bool IgnorePrepare = false;
+        public bool IgnoreSPC = false;
+        public bool FailSPC = false;
+        public bool FailCommit = false;
+        public bool FailRollback = false;
+        public bool UseSingle = false;
+        public Exception ThrowThisException = null;
+
+        private Guid _guid;
+
+        public int Actual
+        {
+            get { return _actual; }
+        }
+
+        public int Value
+        {
+            get { return _transaction == null ? _actual : _tmpValue; }
+            set
+            {
+                if (Transaction.Current == null)
+                {
+                    /* Not in a transaction */
+                    _actual = value;
+                    return;
+                }
+                /* FIXME: Do what in this case? */
+                if (_transaction != null)
+                    Debug.WriteLine("WARNING: Setting value more than once");
+
+                if (_transaction != Transaction.Current)
+                {
+                    _transaction = Transaction.Current;
+
+                    if (Type == ResourceManagerType.Promotable)
+                    {
+                        _transaction.EnlistPromotableSinglePhase(new PromotableSinglePhaseNotification(this));
```

Comment:
Ok, so I should basically add an Assert.True around all calls to enlistPromotableSinglePhase that are supposed to succeed?  Assuming yes, I'll do that, and if any then cause failures, I'll undo it and add a TODO for follow-up.